File: sdk/typescript/src/pluginruntime.ts
import { create } from "@bufbuild/protobuf";
import { EmptySchema } from "@bufbuild/protobuf/wkt";
import {
  Code,
  ConnectError,
  type ServiceImpl,
} from "@connectrpc/connect";

import {
  HostedPluginSchema,
  ListPluginRuntimeSessionsResponseSchema,
  PreparePluginRuntimeWorkspaceResponseSchema,
  PluginRuntimeEgressMode as ProtoPluginRuntimeEgressMode,
  PluginRuntimeProvider as PluginRuntimeProviderService,
  PluginRuntimeSessionSchema,
  PluginRuntimeSupportSchema,
  type GetPluginRuntimeSessionRequest as ProtoGetPluginRuntimeSessionRequest,
  type ListPluginRuntimeSessionsRequest as ProtoListPluginRuntimeSessionsRequest,
  type PreparePluginRuntimeWorkspaceRequest as ProtoPreparePluginRuntimeWorkspaceRequest,
  type RemovePluginRuntimeWorkspaceRequest as ProtoRemovePluginRuntimeWorkspaceRequest,
  type StartHostedPluginRequest as ProtoStartHostedPluginRequest,
  type StartPluginRuntimeSessionRequest as ProtoStartPluginRuntimeSessionRequest,
  type StopPluginRuntimeSessionRequest as ProtoStopPluginRuntimeSessionRequest,
} from "./internal/gen/v1/pluginruntime_pb.ts";
import {
  timestampFromDate,
} from "./protocol.ts";
import { errorMessage, type MaybePromise } from "./api.ts";
import { ProviderBase, type ProviderBaseOptions } from "./provider.ts";

type PluginRuntimeProviderServiceImpl = Partial<
  ServiceImpl<typeof PluginRuntimeProviderService>
>;

/** Native egress-mode constants for authored plugin runtime providers. */
export const PluginRuntimeEgressMode = {
  UNSPECIFIED: ProtoPluginRuntimeEgressMode.UNSPECIFIED,
  NONE: ProtoPluginRuntimeEgressMode.NONE,
  CIDR: ProtoPluginRuntimeEgressMode.CIDR,
  HOSTNAME: ProtoPluginRuntimeEgressMode.HOSTNAME,
} as const;
export type PluginRuntimeEgressMode =
  (typeof PluginRuntimeEgressMode)[keyof typeof PluginRuntimeEgressMode];

export interface PluginRuntimeSupport {
  canHostPlugins?: boolean | undefined;
  egressMode?: PluginRuntimeEgressMode | undefined;
  supportsPrepareWorkspace?: boolean | undefined;
}

export interface PluginRuntimeSessionLifecycle {
  startedAt?: Date | undefined;
  recommendedDrainAt?: Date | undefined;
  expiresAt?: Date | undefined;
}

export interface PluginRuntimeSession {
  id?: string | undefined;
  state?: string | undefined;
  metadata?: Record<string, string> | undefined;
  lifecycle?: PluginRuntimeSessionLifecycle | undefined;
  stateReason?: string | undefined;
  stateMessage?: string | undefined;
}

export interface PluginRuntimeImagePullAuth {
  dockerConfigJson?: string | undefined;
}

export interface StartPluginRuntimeSessionRequest {
  pluginName: string;
  template?: string | undefined;
  image?: string | undefined;
  metadata?: Record<string, string> | undefined;
  imagePullAuth?: PluginRuntimeImagePullAuth | undefined;
}

export interface GetPluginRuntimeSessionRequest {
  sessionId: string;
}

export interface ListPluginRuntimeSessionsRequest {}

export interface StopPluginRuntimeSessionRequest {
  sessionId: string;
}

export interface PluginRuntimeAgentWorkspaceGitCheckout {
  url?: string | undefined;
  ref?: string | undefined;
  path?: string | undefined;
}

export interface PluginRuntimeAgentWorkspace {
  checkouts?: readonly PluginRuntimeAgentWorkspaceGitCheckout[] | undefined;
  cwd?: string | undefined;
}

export interface PluginRuntimePreparedAgentWorkspace {
  root?: string | undefined;
  cwd?: string | undefined;
}

export interface PreparePluginRuntimeWorkspaceRequest {
  sessionId: string;
  agentSessionId: string;
  workspace?: PluginRuntimeAgentWorkspace | undefined;
}

export interface PreparePluginRuntimeWorkspaceResponse {
  workspace?: PluginRuntimePreparedAgentWorkspace | undefined;
}

export interface RemovePluginRuntimeWorkspaceRequest {
  sessionId: string;
  agentSessionId: string;
}

export interface StartHostedPluginRequest {
  sessionId: string;
  pluginName: string;
  command?: string | undefined;
  args?: readonly string[] | undefined;
  env?: Record<string, string> | undefined;
  allowedHosts?: readonly string[] | undefined;
  defaultAction?: string | undefined;
  hostBinary?: string | undefined;
}

export interface HostedPlugin {
  id?: string | undefined;
  sessionId?: string | undefined;
  pluginName?: string | undefined;
  dialTarget?: string | undefined;
}

export interface PluginRuntimeProviderOptions extends ProviderBaseOptions {
  getSupport: () => MaybePromise<PluginRuntimeSupport>;
  startSession: (
    request: StartPluginRuntimeSessionRequest,
  ) => MaybePromise<PluginRuntimeSession>;
  getSession: (
    request: GetPluginRuntimeSessionRequest,
  ) => MaybePromise<PluginRuntimeSession>;
  listSessions: (
    request: ListPluginRuntimeSessionsRequest,
  ) => MaybePromise<readonly PluginRuntimeSession[]>;
  stopSession: (request: StopPluginRuntimeSessionRequest) => MaybePromise<void>;
  prepareWorkspace?: (
    request: PreparePluginRuntimeWorkspaceRequest,
  ) => MaybePromise<PreparePluginRuntimeWorkspaceResponse>;
  removeWorkspace?: (
    request: RemovePluginRuntimeWorkspaceRequest,
  ) => MaybePromise<void>;
  startPlugin: (
    request: StartHostedPluginRequest,
  ) => MaybePromise<HostedPlugin>;
}

export class PluginRuntimeProvider extends ProviderBase {
  readonly kind = "runtime" as const;

  private readonly getSupportHandler: PluginRuntimeProviderOptions["getSupport"];
  private readonly startSessionHandler: PluginRuntimeProviderOptions["startSession"];
  private readonly getSessionHandler: PluginRuntimeProviderOptions["getSession"];
  private readonly listSessionsHandler: PluginRuntimeProviderOptions["listSessions"];
  private readonly stopSessionHandler: PluginRuntimeProviderOptions["stopSession"];
  private readonly prepareWorkspaceHandler: PluginRuntimeProviderOptions["prepareWorkspace"];
  private readonly removeWorkspaceHandler: PluginRuntimeProviderOptions["removeWorkspace"];
  private readonly startPluginHandler: PluginRuntimeProviderOptions["startPlugin"];

  constructor(options: PluginRuntimeProviderOptions) {
    super(options);
    this.getSupportHandler = options.getSupport;
    this.startSessionHandler = options.startSession;
    this.getSessionHandler = options.getSession;
    this.listSessionsHandler = options.listSessions;
    this.stopSessionHandler = options.stopSession;
    this.prepareWorkspaceHandler = options.prepareWorkspace;
    this.removeWorkspaceHandler = options.removeWorkspace;
    this.startPluginHandler = options.startPlugin;
  }

  async getSupport(): Promise<PluginRuntimeSupport> {
    return await this.getSupportHandler();
  }

  async startSession(
    request: StartPluginRuntimeSessionRequest,
  ): Promise<PluginRuntimeSession> {
    return await this.startSessionHandler(request);
  }

  async getSession(
    request: GetPluginRuntimeSessionRequest,
  ): Promise<PluginRuntimeSession> {
    return await this.getSessionHandler(request);
  }

  async listSessions(
    request: ListPluginRuntimeSessionsRequest,
  ): Promise<readonly PluginRuntimeSession[]> {
    return await this.listSessionsHandler(request);
  }

  async stopSession(request: StopPluginRuntimeSessionRequest): Promise<void> {
    await this.stopSessionHandler(request);
  }

  async prepareWorkspace(
    request: PreparePluginRuntimeWorkspaceRequest,
  ): Promise<PreparePluginRuntimeWorkspaceResponse> {
    if (!this.prepareWorkspaceHandler) {
      throw new ConnectError(
        "plugin runtime provider prepare workspace is not implemented",
        Code.Unimplemented,
      );
    }
    return await this.prepareWorkspaceHandler(request);
  }

  async removeWorkspace(
    request: RemovePluginRuntimeWorkspaceRequest,
  ): Promise<void> {
    if (!this.removeWorkspaceHandler) {
      throw new ConnectError(
        "plugin runtime provider remove workspace is not implemented",
        Code.Unimplemented,
      );
    }
    await this.removeWorkspaceHandler(request);
  }

  async startPlugin(
    request: StartHostedPluginRequest,
  ): Promise<HostedPlugin> {
    return await this.startPluginHandler(request);
  }
}

export function definePluginRuntimeProvider(
  options: PluginRuntimeProviderOptions,
): PluginRuntimeProvider {
  return new PluginRuntimeProvider(options);
}

export function isPluginRuntimeProvider(
  value: unknown,
): value is PluginRuntimeProvider {
  return (
    value instanceof PluginRuntimeProvider ||
    (typeof value === "object" &&
      value !== null &&
      "kind" in value &&
      (value as { kind?: unknown }).kind === "runtime" &&
      "getSupport" in value &&
      "startSession" in value &&
      "startPlugin" in value)
  );
}

export function createPluginRuntimeProviderService(
  provider: PluginRuntimeProvider,
): PluginRuntimeProviderServiceImpl {
  return {
    async getSupport() {
      return create(
        PluginRuntimeSupportSchema,
        pluginRuntimeSupportToProto(
          await invokePluginRuntimeProvider("get support", () =>
            provider.getSupport(),
          ),
        ),
      );
    },
    async startSession(request) {
      return create(
        PluginRuntimeSessionSchema,
        pluginRuntimeSessionToProto(
          await invokePluginRuntimeProvider("start session", () =>
            provider.startSession(startSessionRequestFromProto(request)),
          ),
        ),
      );
    },
    async getSession(request) {
      return create(
        PluginRuntimeSessionSchema,
        pluginRuntimeSessionToProto(
          await invokePluginRuntimeProvider("get session", () =>
            provider.getSession(getSessionRequestFromProto(request)),
          ),
        ),
      );
    },
    async listSessions(request) {
      return create(ListPluginRuntimeSessionsResponseSchema, {
        sessions: (
          await invokePluginRuntimeProvider("list sessions", () =>
            provider.listSessions(listSessionsRequestFromProto(request)),
          )
        ).map(pluginRuntimeSessionToProto),
      });
    },
    async stopSession(request) {
      await invokePluginRuntimeProvider("stop session", () =>
        provider.stopSession(stopSessionRequestFromProto(request)),
      );
      return create(EmptySchema);
    },
    async prepareWorkspace(request) {
      return create(
        PreparePluginRuntimeWorkspaceResponseSchema,
        prepareWorkspaceResponseToProto(
          await invokePluginRuntimeProvider("prepare workspace", () =>
            provider.prepareWorkspace(prepareWorkspaceRequestFromProto(request)),
          ),
        ),
      );
    },
    async removeWorkspace(request) {
      await invokePluginRuntimeProvider("remove workspace", () =>
        provider.removeWorkspace(removeWorkspaceRequestFromProto(request)),
      );
      return create(EmptySchema);
    },
    async startPlugin(request) {
      return create(
        HostedPluginSchema,
        hostedPluginToProto(
          await invokePluginRuntimeProvider("start plugin", () =>
            provider.startPlugin(startHostedPluginRequestFromProto(request)),
          ),
        ),
      );
    },
  };
}

function startSessionRequestFromProto(
  request: ProtoStartPluginRuntimeSessionRequest,
): StartPluginRuntimeSessionRequest {
  return {
    pluginName: request.pluginName,
    template: request.template,
    image: request.image,
    metadata: { ...request.metadata },
    imagePullAuth: request.imagePullAuth === undefined
      ? undefined
      : { dockerConfigJson: request.imagePullAuth.dockerConfigJson },
  };
}

function getSessionRequestFromProto(
  request: ProtoGetPluginRuntimeSessionRequest,
): GetPluginRuntimeSessionRequest {
  return { sessionId: request.sessionId };
}

function listSessionsRequestFromProto(
  _request: ProtoListPluginRuntimeSessionsRequest,
): ListPluginRuntimeSessionsRequest {
  return {};
}

function stopSessionRequestFromProto(
  request: ProtoStopPluginRuntimeSessionRequest,
): StopPluginRuntimeSessionRequest {
  return { sessionId: request.sessionId };
}

function prepareWorkspaceRequestFromProto(
  request: ProtoPreparePluginRuntimeWorkspaceRequest,
): PreparePluginRuntimeWorkspaceRequest {
  return {
    sessionId: request.sessionId,
    agentSessionId: request.agentSessionId,
    workspace: request.workspace === undefined
      ? undefined
      : {
          checkouts: request.workspace.checkouts.map((checkout) => ({
            url: checkout.url,
            ref: checkout.ref,
            path: checkout.path,
          })),
          cwd: request.workspace.cwd,
        },
  };
}

function removeWorkspaceRequestFromProto(
  request: ProtoRemovePluginRuntimeWorkspaceRequest,
): RemovePluginRuntimeWorkspaceRequest {
  return {
    sessionId: request.sessionId,
    agentSessionId: request.agentSessionId,
  };
}

function startHostedPluginRequestFromProto(
  request: ProtoStartHostedPluginRequest,
): StartHostedPluginRequest {
  return {
    sessionId: request.sessionId,
    pluginName: request.pluginName,
    command: request.command,
    args: [...request.args],
    env: { ...request.env },
    allowedHosts: [...request.allowedHosts],
    defaultAction: request.defaultAction,
    hostBinary: request.hostBinary,
  };
}

function pluginRuntimeSupportToProto(support: PluginRuntimeSupport) {
  return {
    canHostPlugins: support.canHostPlugins ?? false,
    egressMode: support.egressMode ?? PluginRuntimeEgressMode.UNSPECIFIED,
    supportsPrepareWorkspace: support.supportsPrepareWorkspace ?? false,
  };
}

function pluginRuntimeSessionToProto(session: PluginRuntimeSession) {
  return {
    id: session.id ?? "",
    state: session.state ?? "",
    metadata: { ...(session.metadata ?? {}) },
    lifecycle: session.lifecycle === undefined
      ? undefined
      : {
          startedAt: session.lifecycle.startedAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.startedAt),
          recommendedDrainAt: session.lifecycle.recommendedDrainAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.recommendedDrainAt),
          expiresAt: session.lifecycle.expiresAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.expiresAt),
        },
    stateReason: session.stateReason ?? "",
    stateMessage: session.stateMessage ?? "",
  };
}

function prepareWorkspaceResponseToProto(
  response: PreparePluginRuntimeWorkspaceResponse,
) {
  return {
    workspace: response.workspace === undefined
      ? undefined
      : {
          root: response.workspace.root ?? "",
          cwd: response.workspace.cwd ?? "",
        },
  };
}

function hostedPluginToProto(plugin: HostedPlugin) {
  return {
    id: plugin.id ?? "",
    sessionId: plugin.sessionId ?? "",
    pluginName: plugin.pluginName ?? "",
    dialTarget: plugin.dialTarget ?? "",
  };
}

async function invokePluginRuntimeProvider<T>(
  label: string,
  fn: () => MaybePromise<T>,
): Promise<T> {
  try {
    return await fn();
  } catch (error) {
    if (error instanceof ConnectError) {
      throw error;
    }
    throw new ConnectError(
      `plugin runtime provider ${label}: ${errorMessage(error)}`,
      Code.Unknown,
    );
  }
}

File: sdk/rust/src/plugin_runtime.rs
use std::sync::Arc;
use std::time::SystemTime;

use tonic::codegen::async_trait;
use tonic::{Request as GrpcRequest, Response as GrpcResponse, Status};

use crate::agent::{AgentPreparedWorkspace, AgentWorkspaceInput};
use crate::api::RuntimeMetadata;
use crate::error::Result as ProviderResult;
use crate::generated::v1::{self as pb};
use crate::protocol;
use crate::rpc_status::rpc_status;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
#[repr(i32)]
pub enum PluginRuntimeEgressMode {
    #[default]
    Unspecified = 0,
    None = 1,
    Cidr = 2,
    Hostname = 3,
}

impl PluginRuntimeEgressMode {
    pub const fn as_i32(self) -> i32 {
        self as i32
    }

    pub const fn from_i32_lossy(value: i32) -> Self {
        match value {
            1 => Self::None,
            2 => Self::Cidr,
            3 => Self::Hostname,
            _ => Self::Unspecified,
        }
    }
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct PluginRuntimeSupport {
    pub can_host_plugins: bool,
    pub egress_mode: PluginRuntimeEgressMode,
    pub supports_prepare_workspace: bool,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PluginRuntimeSession {
    pub id: String,
    pub state: String,
    pub metadata: std::collections::BTreeMap<String, String>,
    pub lifecycle: Option<PluginRuntimeSessionLifecycle>,
    pub state_reason: String,
    pub state_message: String,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct PluginRuntimeSessionLifecycle {
    pub started_at: Option<SystemTime>,
    pub recommended_drain_at: Option<SystemTime>,
    pub expires_at: Option<SystemTime>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct PluginRuntimeImagePullAuth {
    pub docker_config_json: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct StartPluginRuntimeSessionRequest {
    pub plugin_name: String,
    pub template: String,
    pub image: String,
    pub metadata: std::collections::BTreeMap<String, String>,
    pub image_pull_auth: Option<PluginRuntimeImagePullAuth>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct GetPluginRuntimeSessionRequest {
    pub session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ListPluginRuntimeSessionsRequest {}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ListPluginRuntimeSessionsResponse {
    pub sessions: Vec<PluginRuntimeSession>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct StopPluginRuntimeSessionRequest {
    pub session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PreparePluginRuntimeWorkspaceRequest {
    pub session_id: String,
    pub agent_session_id: String,
    pub workspace: Option<AgentWorkspaceInput>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct PreparePluginRuntimeWorkspaceResponse {
    pub workspace: Option<AgentPreparedWorkspace>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct RemovePluginRuntimeWorkspaceRequest {
    pub session_id: String,
    pub agent_session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct StartHostedPluginRequest {
    pub session_id: String,
    pub plugin_name: String,
    pub command: String,
    pub args: Vec<String>,
    pub env: std::collections::BTreeMap<String, String>,
    pub allowed_hosts: Vec<String>,
    pub default_action: String,
    pub host_binary: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct HostedPlugin {
    pub id: String,
    pub session_id: String,
    pub plugin_name: String,
    pub dial_target: String,
}

fn support_to_proto(value: PluginRuntimeSupport) -> pb::PluginRuntimeSupport {
    pb::PluginRuntimeSupport {
        can_host_plugins: value.can_host_plugins,
        egress_mode: value.egress_mode.as_i32(),
        supports_prepare_workspace: value.supports_prepare_workspace,
    }
}

fn session_to_proto(value: PluginRuntimeSession) -> pb::PluginRuntimeSession {
    pb::PluginRuntimeSession {
        id: value.id,
        state: value.state,
        metadata: value.metadata,
        lifecycle: value
            .lifecycle
            .map(|lifecycle| pb::PluginRuntimeSessionLifecycle {
                started_at: lifecycle
                    .started_at
                    .map(protocol::timestamp_from_system_time),
                recommended_drain_at: lifecycle
                    .recommended_drain_at
                    .map(protocol::timestamp_from_system_time),
                expires_at: lifecycle
                    .expires_at
                    .map(protocol::timestamp_from_system_time),
            }),
        state_reason: value.state_reason,
        state_message: value.state_message,
    }
}

fn start_session_request_from_proto(
    value: pb::StartPluginRuntimeSessionRequest,
) -> StartPluginRuntimeSessionRequest {
    StartPluginRuntimeSessionRequest {
        plugin_name: value.plugin_name,
        template: value.template,
        image: value.image,
        metadata: value.metadata,
        image_pull_auth: value
            .image_pull_auth
            .map(|auth| PluginRuntimeImagePullAuth {
                docker_config_json: auth.docker_config_json,
            }),
    }
}

fn list_sessions_response_to_proto(
    value: ListPluginRuntimeSessionsResponse,
) -> pb::ListPluginRuntimeSessionsResponse {
    pb::ListPluginRuntimeSessionsResponse {
        sessions: value.sessions.into_iter().map(session_to_proto).collect(),
    }
}

fn prepare_workspace_request_from_proto(
    value: pb::PreparePluginRuntimeWorkspaceRequest,
) -> PreparePluginRuntimeWorkspaceRequest {
    PreparePluginRuntimeWorkspaceRequest {
        session_id: value.session_id,
        agent_session_id: value.agent_session_id,
        workspace: value.workspace.map(|workspace| AgentWorkspaceInput {
            checkouts: workspace
                .checkouts
                .into_iter()
                .map(|checkout| crate::agent::AgentWorkspaceGitCheckoutInput {
                    url: checkout.url,
                    reference: checkout.r#ref,
                    path: checkout.path,
                })
                .collect(),
            cwd: workspace.cwd,
        }),
    }
}

fn prepare_workspace_response_to_proto(
    value: PreparePluginRuntimeWorkspaceResponse,
) -> pb::PreparePluginRuntimeWorkspaceResponse {
    pb::PreparePluginRuntimeWorkspaceResponse {
        workspace: value.workspace.map(|workspace| pb::PreparedAgentWorkspace {
            root: workspace.root,
            cwd: workspace.cwd,
        }),
    }
}

fn start_plugin_request_from_proto(
    value: pb::StartHostedPluginRequest,
) -> StartHostedPluginRequest {
    StartHostedPluginRequest {
        session_id: value.session_id,
        plugin_name: value.plugin_name,
        command: value.command,
        args: value.args,
        env: value.env,
        allowed_hosts: value.allowed_hosts,
        default_action: value.default_action,
        host_binary: value.host_binary,
    }
}

fn hosted_plugin_to_proto(value: HostedPlugin) -> pb::HostedPlugin {
    pb::HostedPlugin {
        id: value.id,
        session_id: value.session_id,
        plugin_name: value.plugin_name,
        dial_target: value.dial_target,
    }
}

#[async_trait]
/// Provider trait for serving hosted plugin-runtime sessions.
pub trait PluginRuntimeProvider: Send + Sync + 'static {
    /// Configures the provider before it starts serving requests.
    async fn configure(
        &self,
        _name: &str,
        _config: serde_json::Map<String, serde_json::Value>,
    ) -> ProviderResult<()> {
        Ok(())
    }

    /// Returns runtime metadata that should augment the static manifest.
    fn metadata(&self) -> Option<RuntimeMetadata> {
        None
    }

    /// Returns non-fatal warnings the host should surface to users.
    fn warnings(&self) -> Vec<String> {
        Vec::new()
    }

    /// Performs an optional health check.
    async fn health_check(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Starts provider-owned background work after configuration.
    async fn start(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Shuts the provider down before the runtime exits.
    async fn close(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Returns the runtime capabilities supported by this provider.
    async fn get_support(&self, _request: ()) -> ProviderResult<PluginRuntimeSupport> {
        Err(crate::Error::unimplemented(
            "runtime get support is not implemented",
        ))
    }

    /// Starts a hosted plugin-runtime session.
    async fn start_session(
        &self,
        _request: StartPluginRuntimeSessionRequest,
    ) -> ProviderResult<PluginRuntimeSession> {
        Err(crate::Error::unimplemented(
            "runtime start session is not implemented",
        ))
    }

    /// Returns one hosted plugin-runtime session by ID.
    async fn get_session(
        &self,
        _request: GetPluginRuntimeSessionRequest,
    ) -> ProviderResult<PluginRuntimeSession> {
        Err(crate::Error::unimplemented(
            "runtime get session is not implemented",
        ))
    }

    /// Lists hosted plugin-runtime sessions.
    async fn list_sessions(
        &self,
        _request: ListPluginRuntimeSessionsRequest,
    ) -> ProviderResult<ListPluginRuntimeSessionsResponse> {
        Err(crate::Error::unimplemented(
            "runtime list sessions is not implemented",
        ))
    }

    /// Stops a hosted plugin-runtime session.
    async fn stop_session(&self, _request: StopPluginRuntimeSessionRequest) -> ProviderResult<()> {
        Err(crate::Error::unimplemented(
            "runtime stop session is not implemented",
        ))
    }

    /// Prepares an agent workspace for use by a hosted plugin.
    async fn prepare_workspace(
        &self,
        _request: PreparePluginRuntimeWorkspaceRequest,
    ) -> ProviderResult<PreparePluginRuntimeWorkspaceResponse> {
        Err(crate::Error::unimplemented(
            "runtime prepare workspace is not implemented",
        ))
    }

    /// Removes a previously prepared agent workspace.
    async fn remove_workspace(
        &self,
        _request: RemovePluginRuntimeWorkspaceRequest,
    ) -> ProviderResult<()> {
        Err(crate::Error::unimplemented(
            "runtime remove workspace is not implemented",
        ))
    }

    /// Starts one hosted plugin process inside a runtime session.
    async fn start_plugin(
        &self,
        _request: StartHostedPluginRequest,
    ) -> ProviderResult<HostedPlugin> {
        Err(crate::Error::unimplemented(
            "runtime start plugin is not implemented",
        ))
    }
}

#[derive(Clone)]
pub(crate) struct PluginRuntimeServer<P> {
    provider: Arc<P>,
}

impl<P> PluginRuntimeServer<P> {
    pub(crate) fn new(provider: Arc<P>) -> Self {
        Self { provider }
    }
}

#[async_trait]
impl<P> pb::plugin_runtime_provider_server::PluginRuntimeProvider for PluginRuntimeServer<P>
where
    P: PluginRuntimeProvider,
{
    async fn get_support(
        &self,
        request: GrpcRequest<()>,
    ) -> std::result::Result<GrpcResponse<pb::PluginRuntimeSupport>, Status> {
        let support = self
            .provider
            .get_support(request.into_inner())
            .await
            .map_err(|error| rpc_status("runtime get support", error))?;
        Ok(GrpcResponse::new(support_to_proto(support)))
    }

    async fn start_session(
        &self,
        request: GrpcRequest<pb::StartPluginRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<pb::PluginRuntimeSession>, Status> {
        let session = self
            .provider
            .start_session(start_session_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime start session", error))?;
        Ok(GrpcResponse::new(session_to_proto(session)))
    }

    async fn get_session(
        &self,
        request: GrpcRequest<pb::GetPluginRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<pb::PluginRuntimeSession>, Status> {
        let session = self
            .provider
            .get_session({
                let request = request.into_inner();
                GetPluginRuntimeSessionRequest {
                    session_id: request.session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime get session", error))?;
        Ok(GrpcResponse::new(session_to_proto(session)))
    }

    async fn list_sessions(
        &self,
        request: GrpcRequest<pb::ListPluginRuntimeSessionsRequest>,
    ) -> std::result::Result<GrpcResponse<pb::ListPluginRuntimeSessionsResponse>, Status> {
        let response = self
            .provider
            .list_sessions({
                let _request = request.into_inner();
                ListPluginRuntimeSessionsRequest {}
            })
            .await
            .map_err(|error| rpc_status("runtime list sessions", error))?;
        Ok(GrpcResponse::new(list_sessions_response_to_proto(response)))
    }

    async fn stop_session(
        &self,
        request: GrpcRequest<pb::StopPluginRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.provider
            .stop_session({
                let request = request.into_inner();
                StopPluginRuntimeSessionRequest {
                    session_id: request.session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime stop session", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn prepare_workspace(
        &self,
        request: GrpcRequest<pb::PreparePluginRuntimeWorkspaceRequest>,
    ) -> std::result::Result<GrpcResponse<pb::PreparePluginRuntimeWorkspaceResponse>, Status> {
        let response = self
            .provider
            .prepare_workspace(prepare_workspace_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime prepare workspace", error))?;
        Ok(GrpcResponse::new(prepare_workspace_response_to_proto(
            response,
        )))
    }

    async fn remove_workspace(
        &self,
        request: GrpcRequest<pb::RemovePluginRuntimeWorkspaceRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.provider
            .remove_workspace({
                let request = request.into_inner();
                RemovePluginRuntimeWorkspaceRequest {
                    session_id: request.session_id,
                    agent_session_id: request.agent_session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime remove workspace", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn start_plugin(
        &self,
        request: GrpcRequest<pb::StartHostedPluginRequest>,
    ) -> std::result::Result<GrpcResponse<pb::HostedPlugin>, Status> {
        let plugin = self
            .provider
            .start_plugin(start_plugin_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime start plugin", error))?;
        Ok(GrpcResponse::new(hosted_plugin_to_proto(plugin)))
    }
}
